File: src/core/xml-parser.ts
import { DOMParser } from 'xmldom';
import * as xpath from 'xpath';
import * as fs from 'fs';
import { XmlFile } from '../types';

export class XmlParser {
  private domParser: DOMParser;

  constructor() {
    this.domParser = new DOMParser();
  }

  parseXmlFile(filePath: string): XmlFile {
    try {
      const xmlContent = fs.readFileSync(filePath, 'utf-8');
      const document = this.domParser.parseFromString(xmlContent, 'text/xml');
      
      if (this.hasParseErrors(document)) {
        throw new Error(`Invalid XML in file: ${filePath}`);
      }

      return {
        path: filePath,
        content: document
      };
    } catch (error) {
      throw new Error(`Failed to parse XML file ${filePath}: ${error instanceof Error ? error.message : String(error)}`);
    }
  }

  parseXmlFiles(filePaths: string[]): XmlFile[] {
    return filePaths.map(filePath => this.parseXmlFile(filePath));
  }

  evaluateXPath(document: Document, xpathExpression: string): string | null {
    try {
      const result = xpath.select(xpathExpression, document);
      
      if (Array.isArray(result) && result.length > 0) {
        const node = result[0];
        if (typeof node === 'string') {
          return node;
        }
        if (node && typeof node === 'object' && 'nodeValue' in node) {
          return (node as any).nodeValue || (node as any).textContent || null;
        }
        if (node && typeof node === 'object' && 'textContent' in node) {
          return (node as any).textContent || null;
        }
      }
      
      return null;
    } catch (error) {
      throw new Error(`XPath evaluation failed for expression "${xpathExpression}": ${error instanceof Error ? error.message : String(error)}`);
    }
  }

  private hasParseErrors(document: Document): boolean {
    const parseErrors = document.getElementsByTagName('parsererror');
    return parseErrors.length > 0;
  }
}

File: xml_inspector/core/xml_parser.py
"""XML parsing and XPath evaluation functionality."""

from typing import List, Optional, Union
from pathlib import Path
import logging

from lxml import etree

from ..types import XmlFile

logger = logging.getLogger(__name__)


class XmlParseError(Exception):
    """Raised when XML parsing fails."""
    pass


class XPathEvaluationError(Exception):
    """Raised when XPath evaluation fails."""
    pass


class XmlParser:
    """Handles XML file parsing and XPath evaluation."""
    
    def __init__(self) -> None:
        """Initialize the XML parser."""
        self.parser = etree.XMLParser(recover=False, strip_cdata=False)
    
    def parse_xml_file(self, file_path: Union[str, Path]) -> XmlFile:
        """
        Parse a single XML file.
        
        Args:
            file_path: Path to the XML file
            
        Returns:
            XmlFile object containing the parsed content
            
        Raises:
            XmlParseError: If the file cannot be parsed
        """
        file_path = Path(file_path)
        
        try:
            if not file_path.exists():
                raise XmlParseError(f"File not found: {file_path}")
            
            if not file_path.is_file():
                raise XmlParseError(f"Path is not a file: {file_path}")
            
            with open(file_path, 'r', encoding='utf-8') as f:
                xml_content = f.read()
            
            try:
                document = etree.fromstring(xml_content.encode('utf-8'), self.parser)
            except etree.XMLSyntaxError as e:
                raise XmlParseError(f"Invalid XML in file {file_path}: {e}")
            
            return XmlFile(path=str(file_path), content=document)
            
        except Exception as e:
            if isinstance(e, XmlParseError):
                raise
            raise XmlParseError(f"Failed to parse XML file {file_path}: {e}")
    
    def parse_xml_files(self, file_paths: List[Union[str, Path]]) -> List[XmlFile]:
        """
        Parse multiple XML files.
        
        Args:
            file_paths: List of paths to XML files
            
        Returns:
            List of XmlFile objects
        """
        return [self.parse_xml_file(file_path) for file_path in file_paths]
    
    def evaluate_xpath(self, document: etree._Element, xpath_expression: str) -> Optional[str]:
        """
        Evaluate an XPath expression against an XML document.
        
        Args:
            document: The XML document (lxml Element)
            xpath_expression: XPath expression to evaluate
            
        Returns:
            String result of the XPath evaluation, or None if no match
            
        Raises:
            XPathEvaluationError: If XPath evaluation fails
        """
        try:
            result = document.xpath(xpath_expression)
            
            if not result:
                return None
            
            # Handle different types of XPath results
            if isinstance(result, list) and len(result) > 0:
                first_result = result[0]
                
                # Text node
                if isinstance(first_result, str):
                    return first_result.strip()
                
                # Element node - get text content
                if hasattr(first_result, 'text') and first_result.text is not None:
                    return first_result.text.strip()
                
                # Attribute node
                if hasattr(first_result, 'strip'):
                    return first_result.strip()
                
                # Element node with no direct text - get all text content
                if hasattr(first_result, 'itertext'):
                    text_content = ''.join(first_result.itertext()).strip()
                    return text_content if text_content else None
            
            return None
            
        except etree.XPathEvalError as e:
            raise XPathEvaluationError(f'XPath evaluation failed for expression "{xpath_expression}": {e}')
        except Exception as e:
            raise XPathEvaluationError(f'Unexpected error evaluating XPath "{xpath_expression}": {e}')